classes/village/VillageUpgradeConfig.php

```diff
@@ -0,0 +1,2188 @@
+    const UPGRADE_KEY_INSPIRED_ITAMAE_I = 'INSPIRED_ITAMAE_I';
+    const UPGRADE_KEY_LUCK_IN_LEFTOVERS_I = 'LUCK_IN_LEFTOVERS_I';
+    const UPGRADE_KEY_LUCK_IN_LEFTOVERS_II = 'LUCK_IN_LEFTOVERS_II';
+
+    /* constant used to identify individual effects that may be present in multiple upgrades */
+    const UPGRADE_EFFECT_MATERIALS_PRODUCTION = 'MATERIALS_PRODUCTION';
+    const UPGRADE_EFFECT_FOOD_PRODUCTION = 'FOOD_PRODUCTION';
+    const UPGRADE_EFFECT_WEALTH_PRODUCTION = 'WEALTH_PRODUCTION';
+    const UPGRADE_EFFECT_CONSTRUCTION_SPEED = 'CONSTRUCTION_SPEED';
+    const UPGRADE_EFFECT_RESEARCH_SPEED = 'RESEARCH_SPEED';
+    const UPGRADE_EFFECT_RESEARCH_T1_ENABLED = 'RESEARCH_T1';
+    const UPGRADE_EFFECT_RESEARCH_T2_ENABLED = 'RESEARCH_T2_ENABLED';
+    const UPGRADE_EFFECT_RESEARCH_T3_ENABLED = 'RESEARCH_T3_ENABLED';
+    const UPGRADE_EFFECT_CONSTRUCTION_T1_ENABLED = 'CONSTRUCTION_T1_ENABLED';
+    const UPGRADE_EFFECT_CONSTRUCTION_T2_ENABLED = 'CONSTRUCTION_T2_ENABLED';
+    const UPGRADE_EFFECT_CONSTRUCTION_T3_ENABLED = 'CONSTRUCTION_T3_ENABLED';
+    const UPGRADE_EFFECT_UPGRADE_UPKEEP = 'UPGRADE_UPKEEP';
+    const UPGRADE_EFFECT_BASE_STABILITY = 'BASE_STABILITY';
+    const UPGRADE_EFFECT_MAX_STABILITY = 'MAX_STABILITY';
+    const UPGRADE_EFFECT_TRAINING_SPEED = 'TRAINING_SPEED';
+    const UPGRADE_EFFECT_DOUBLE_BATTLE_XP_CHANCE = 'DOUBLE_BATTLE_XP_CHANCE';
+    const UPGRADE_EFFECT_VILLAGE_REGEN = 'VILLAGE_REGEN';
+    const UPGRADE_EFFECT_HEAL_ITEM_COST = 'HEAL_ITEM_COST';
+    const UPGRADE_EFFECT_WAR_ACTION_COST = 'WAR_ACTION_COST';
+    const UPGRADE_EFFECT_RAID_SPEED = 'RAID_SPEED';
+    const UPGRADE_EFFECT_RAID_DAMAGE = 'RAID_DAMAGE';
+    const UPGRADE_EFFECT_INFILTRATE_SPEED = 'INFILTRATE_SPEED';
+    const UPGRADE_EFFECT_OCCUPIED_BASE_STABILITY_REDUCTION = 'OCCUPIED_BASE_STABILITY';
+    const UPGRADE_EFFECT_OCCUPIED_MAX_STABILITY_REDUCTION = 'OCCUPIED_MAX_STABILITY';
+    const UPGRADE_EFFECT_PATROL_TIER_CHANCE = 'PATROL_TIER_CHANCE';
+    const UPGRADE_EFFECT_REINFORCE_SPEED = 'REINFORCE_SPEED';
+    const UPGRADE_EFFECT_REINFORCE_HEAL = 'REINFORCE_HEAL';
+    const UPGRADE_EFFECT_CASTLE_HP = 'CASTLE_HP';
+    const UPGRADE_EFFECT_TOWN_HP = 'TOWN_HP';
+    const UPGRADE_EFFECT_RESOURCE_CAPACITY = 'RESOURCE_CAPACITY';
+    const UPGRADE_EFFECT_BLOODLINE_CHANCE = 'BLOODLINE_CHANCE';
+    const UPGRADE_EFFECT_DOUBLE_BATTLE_YEN_CHANCE = 'DOUBLE_BATTLE_YEN_CHANCE';
+    const UPGRADE_EFFECT_RAMEN_SET_ONE = 'RAMEN_SET_ONE';
+    const UPGRADE_EFFECT_RAMEN_DURATION = 'RAMEN_DURATION';
+    const UPGRADE_EFFECT_RAMEN_COST = 'RAMEN_COST';
+    const UPGRADE_EFFECT_MYSTERY_RAMEN_ENABLED = 'MYSTERY_RAMEN_ENABLED';
+    const UPGRADE_EFFECT_MYSTERY_RAMEN_CHANCE = 'MYSTERY_RAMEN_CHANCE';
+
+    /* research requirement types */
+    const UPGRADE_REQUIREMENT_BUILDINGS = "BUILDINGS";
+    const UPGRADE_REQUIREMENT_UPGRADES = "UPGRADES";
+
+    /* display names for each upgrade indexed by upgrade key */
+    const UPGRADE_NAMES = [
```

Comment:
Rather than having these as separate arrays, could we pull the values for each into a DTO and organize them by key? e.g.

```php

class VillageUpgradeConfig {
  /** @var VillageUpgrade[] */
  public static array $upgrades;

  // ...
}

VillageUpgradeConfig::$upgrades = [
  VillageUpgradeConfig::UPGRADE_KEY_RESEARCH_I => new VillageUpgrade(
    name: 'Research I',
    description: 'Allows researching Tier I Upgrades',
    research_costs: [
      WarManager::RESOURCE_MATERIALS => 0,
      WarManager::RESOURCE_FOOD => 0,
      WarManager::RESOURCE_WEALTH => 0,
    ],
    research_time: 0,
    upkeep_costs: [
      WarManager::RESOURCE_MATERIALS => 0,
      WarManager::RESOURCE_FOOD => 0,
      WarManager::RESOURCE_WEALTH => 0,
    ],
    building_tiers_required: [
      VillageBuildingConfig::BUILDING_ACADEMY => 1
    ],
    upgrades_required: [],
    effects: [
      VillageUpgradeConfig::UPGRADE_EFFECT_RESEARCH_T1_ENABLED => 1
    ]
  ),
  // ...
];
```